source/agora/consensus/validation/Transaction.d

```diff
@@ -95,6 +98,9 @@ public string isInvalidReason (
 
         if (sum_unspent.integral() < Amount.MinFreezeAmount.integral())
             return "Transaction: available when the amount is at least 40,000 BOA";
+
+        auto res = checkPayload(tx, sum_unspent);
```

Comment:
I think here we should disallow a data payload if the transaction is a freeze transaction. So similar to how you did it below:

```D
if (tx.payload.data.length != 0)
    return "Transaction: Frozen UTXO cannot be used for data payload";
```

Also, could you move this check more up? If we check this before checking the `sum_unspent`, it will be more efficient.

Reply:
I think that's good, too.